File: libpretixsync/src/main/java/eu/pretix/libpretixsync/check/TicketCheckProvider.java
package eu.pretix.libpretixsync.check;


import java.util.Date;
import java.util.List;
import java.util.Objects;

import eu.pretix.libpretixsync.SentryInterface;
import eu.pretix.libpretixsync.db.Question;
import org.json.JSONObject;

public interface TicketCheckProvider {

    class RequiredAnswer {
        private Question question;
        private String current_value;

        public RequiredAnswer(Question question, String current_value) {
            this.question = question;
            this.current_value = current_value;
        }

        public Question getQuestion() {
            return question;
        }

        public String getCurrentValue() {
            return current_value;
        }
    }

    class Answer {
        private Question question;
        private String value;

        public Answer(Question question, String value) {
            this.question = question;
            this.value = value;
        }

        public String getValue() {
            return value;
        }

        public void setValue(String value) {
            this.value = value;
        }

        public Question getQuestion() {
            return question;
        }
    }

    class CheckResult {
        public enum Type {
            INVALID, VALID, USED, ERROR, UNPAID, PRODUCT, ANSWERS_REQUIRED
        }

        private Type type;
        private String ticket;
        private String variation;
        private String attendee_name;
        private String message;
        private String order_code;
        private Date first_scanned;
        private String addon_text;
        private boolean require_attention;
        private boolean checkin_allowed;
        private List<RequiredAnswer> required_answers;
        private JSONObject position;

        public CheckResult(Type type, String message) {
            this.type = type;
            this.message = message;
        }

        public CheckResult(Type type) {
            this.type = type;
        }

        public Type getType() {
            return type;
        }

        public void setType(Type type) {
            this.type = type;
        }

        public String getTicket() {
            return ticket;
        }

        public void setTicket(String ticket) {
            this.ticket = ticket;
        }

        public String getVariation() {
            return variation;
        }

        public void setVariation(String variation) {
            this.variation = variation;
        }

        public String getAttendee_name() {
            return attendee_name;
        }

        public void setAttendee_name(String attendee_name) {
            this.attendee_name = attendee_name;
        }

        public String getMessage() {
            return message;
        }

        public void setMessage(String message) {
            this.message = message;
        }

        public String getOrderCode() {
            return order_code;
        }

        public void setOrderCode(String order_code) {
            this.order_code = order_code;
        }

        public String getAddonText() {
            return addon_text;
        }

        public void setAddonText(String addon_text) {
            this.addon_text = addon_text;
        }

        public boolean isRequireAttention() {
            return require_attention;
        }

        public void setRequireAttention(boolean require_attention) {
            this.require_attention = require_attention;
        }

        public List<RequiredAnswer> getRequiredAnswers() {
            return required_answers;
        }

        public void setRequiredAnswers(List<RequiredAnswer> required_answers) {
            this.required_answers = required_answers;
        }

        public boolean isCheckinAllowed() {
            return checkin_allowed;
        }

        public void setCheckinAllowed(boolean checkin_allowed) {
            this.checkin_allowed = checkin_allowed;
        }

        public Date getFirstScanned() {
            return first_scanned;
        }

        public void setFirstScanned(Date first_scanned) {
            this.first_scanned = first_scanned;
        }

        public JSONObject getPosition() {
            return position;
        }

        public void setPosition(JSONObject position) {
            this.position = position;
        }
    }

    class SearchResult {

        private String secret;
        private String ticket;
        private String variation;
        private String attendee_name;
        private String order_code;
        private String addon_text;
        private boolean paid;
        private boolean redeemed;
        private boolean require_attention;

        public SearchResult() {
        }

        public SearchResult(SearchResult r) {
            secret = r.secret;
            ticket = r.ticket;
            variation = r.variation;
            attendee_name = r.attendee_name;
            order_code = r.order_code;
            paid = r.paid;
            redeemed = r.redeemed;
            require_attention = r.require_attention;
            addon_text = r.addon_text;
        }

        public String getSecret() {
            return secret;
        }

        public void setSecret(String secret) {
            this.secret = secret;
        }

        public boolean isPaid() {
            return paid;
        }

        public void setPaid(boolean paid) {
            this.paid = paid;
        }

        public boolean isRedeemed() {
            return redeemed;
        }

        public void setRedeemed(boolean redeemed) {
            this.redeemed = redeemed;
        }

        public String getTicket() {
            return ticket;
        }

        public void setTicket(String ticket) {
            this.ticket = ticket;
        }

        public String getVariation() {
            return variation;
        }

        public void setVariation(String variation) {
            this.variation = variation;
        }

        public String getAttendee_name() {
            return attendee_name;
        }

        public void setAttendee_name(String attendee_name) {
            this.attendee_name = attendee_name;
        }

        public String getOrderCode() {
            return order_code;
        }

        public void setOrderCode(String order_code) {
            this.order_code = order_code;
        }

        public String getAddonText() {
            return addon_text;
        }

        public void setAddonText(String addon_text) {
            this.addon_text = addon_text;
        }

        public boolean isRequireAttention() {
            return require_attention;
        }

        public void setRequireAttention(boolean require_attention) {
            this.require_attention = require_attention;
        }
    }

    class StatusResultItemVariation {
        private long id;
        private String name;
        private int checkins;
        private int total;

        public StatusResultItemVariation(long id, String name, int total, int checkins) {
            this.name = name;
            this.checkins = checkins;
            this.total = total;
            this.id = id;
        }

        public long getId() {
            return id;
        }

        public String getName() {
            return name;
        }

        public int getCheckins() {
            return checkins;
        }

        public int getTotal() {
            return total;
        }

        public void setCheckins(int checkins) {
            this.checkins = checkins;
        }

        public void setTotal(int total) {
            this.total = total;
        }
    }

    class StatusResultItem {
        private long id;
        private String name;
        private int checkins;
        private int total;
        private List<StatusResultItemVariation> variations;
        private boolean admission;

        public StatusResultItem(long id, String name, int total, int checkins, List<StatusResultItemVariation> variations, boolean admission) {
            this.name = name;
            this.checkins = checkins;
            this.total = total;
            this.variations = variations;
            this.admission = admission;
            this.id = id;
        }

        public long getId() {
            return id;
        }

        public String getName() {
            return name;
        }

        public int getCheckins() {
            return checkins;
        }

        public int getTotal() {
            return total;
        }

        public boolean isAdmission() {
            return admission;
        }

        public List<StatusResultItemVariation> getVariations() {
            return variations;
        }

        public void setName(String name) {
            this.name = name;
        }

        public void setCheckins(int checkins) {
            this.checkins = checkins;
        }

        public void setTotal(int total) {
            this.total = total;
        }
    }

    class StatusResult {
        private String eventName;
        private int totalTickets;
        private int alreadyScanned;
        private List<StatusResultItem> items;

        public StatusResult(String eventName, int totalTickets, int alreadyScanned, List<StatusResultItem> items) {
            this.eventName = eventName;
            this.totalTickets = totalTickets;
            this.alreadyScanned = alreadyScanned;
            this.items = items;
        }

        public String getEventName() {
            return eventName;
        }

        public int getTotalTickets() {
            return totalTickets;
        }

        public int getAlreadyScanned() {
            return alreadyScanned;
        }

        public List<StatusResultItem> getItems() {
            return items;
        }

        public void setTotalTickets(int totalTickets) {
            this.totalTickets = totalTickets;
        }

        public void setAlreadyScanned(int alreadyScanned) {
            this.alreadyScanned = alreadyScanned;
        }
    }

    CheckResult check(String ticketid, List<Answer> answers, boolean ignore_unpaid, boolean with_badge_data);

    CheckResult check(String ticketid);

    List<SearchResult> search(String query, int page) throws CheckException;

    StatusResult status() throws CheckException;

    public void setSentry(SentryInterface sentry);
}

File: libpretixsync/src/main/java/eu/pretix/libpretixsync/check/TicketCheckProvider.kt
package eu.pretix.libpretixsync.check

import eu.pretix.libpretixsync.SentryInterface
import eu.pretix.libpretixsync.db.Answer
import eu.pretix.libpretixsync.models.db.toModel
import eu.pretix.libpretixsync.sqldelight.Question
import eu.pretix.libpretixsync.models.Question as QuestionModel
import org.json.JSONObject
import java.util.*

interface TicketCheckProvider {
    enum class CheckInType {
        ENTRY, EXIT
    }

    // Old clients expect the requery models on the API
    // This class mimics the relevant fields
    // TODO: List affected versions?
    data class QuestionOutput(
        val server_id: Long,
        val position: Long,
        val required: Boolean,
        val json_data: String,
    ) {
        constructor(model: QuestionModel, jsonData: String) : this(
            server_id = model.serverId,
            required = model.required,
            position = model.position,
            json_data = jsonData,
        )

        fun toModel() = Question(
            server_id = server_id,
            position = position,
            required = required,
            json_data = json_data,
            id = -1L,
            event_slug = null,
        ).toModel()
    }

    class QuestionAnswer {
        private lateinit var _question: QuestionModel
        private lateinit var _jsonData: String

        var currentValue: String? = null

        val question: QuestionOutput
            get() = QuestionOutput(_question, _jsonData)

        constructor(question: QuestionModel, jsonData: String, currentValue: String?) {
            this._question = question
            this._jsonData = jsonData
            this.currentValue = currentValue
        }

        // required for de-serialization
        constructor() {}
    }

    class CheckResult {
        enum class Type {
            INVALID, VALID, USED, ERROR, UNPAID, BLOCKED, INVALID_TIME, CANCELED, PRODUCT, RULES,
            ANSWERS_REQUIRED, AMBIGUOUS, REVOKED, UNAPPROVED
        }

        var type: Type? = null
        var scanType: CheckInType = CheckInType.ENTRY
        var ticket: String? = null
        var variation: String? = null
        var attendee_name: String? = null
        var seat: String? = null
        var message: String? = null
        var orderCode: String? = null
        var positionId: Long? = null
        var firstScanned: Date? = null
        var addonText: String? = null
        var reasonExplanation: String? = null
        var checkinTexts: List<String>? = null
        var isRequireAttention = false
        var isCheckinAllowed = false
        var requiredAnswers: List<QuestionAnswer>? = null
        var shownAnswers: List<QuestionAnswer>? = null
        var position: JSONObject? = null
        var eventSlug: String? = null
        var offline: Boolean = false

        constructor(type: Type?, message: String?, offline: Boolean = false) {
            this.type = type
            this.message = message
            this.offline = offline
        }

        constructor(type: Type?, offline: Boolean = false) {
            this.type = type
            this.offline = offline
        }

        constructor() {  // required for de-serialization
        }

        fun orderCodeAndPositionId(): String? {
            if (orderCode != null && positionId != null && positionId!! > 0) {
                return "${orderCode}-${positionId}"
            } else {
                return orderCode
            }
        }
    }

    class SearchResult {
        enum class Status {
            PAID, CANCELED, PENDING
        }

        var secret: String? = null
        var ticket: String? = null
        var variation: String? = null
        var attendee_name: String? = null
        var seat: String? = null
        var orderCode: String? = null
        var positionId: Long? = null
        var addonText: String? = null
        var status: Status? = null
        var isRedeemed = false
        var isRequireAttention = false
        var position: JSONObject? = null

        constructor() {}
        constructor(r: SearchResult) {
            secret = r.secret
            ticket = r.ticket
            variation = r.variation
            attendee_name = r.attendee_name
            seat = r.seat
            orderCode = r.orderCode
            positionId = r.positionId
            status = r.status
            isRedeemed = r.isRedeemed
            isRequireAttention = r.isRequireAttention
            addonText = r.addonText
            position = r.position
        }

        fun orderCodeAndPositionId(): String? {
            if (orderCode != null && positionId != null && positionId!! > 0) {
                return "${orderCode}-${positionId}"
            } else {
                return orderCode
            }
        }
    }

    class StatusResultItemVariation(var id: Long, var name: String?, var total: Int, var checkins: Int) {
    }

    class StatusResultItem(var id: Long, var name: String?, var total: Int, var checkins: Int, var variations: List<StatusResultItemVariation>?, admission: Boolean) {
        var isAdmission = admission
    }

    class StatusResult(var eventName: String?, var totalTickets: Int, var alreadyScanned: Int, var currentlyInside: Int?, var items: List<StatusResultItem>?) {
    }

    fun check(eventsAndCheckinLists: Map<String, Long>, ticketid: String, source_type: String, answers: List<Answer>?, ignore_unpaid: Boolean, with_badge_data: Boolean, type: CheckInType, nonce: String? = null, allowQuestions: Boolean = true): CheckResult
    fun check(eventsAndCheckinLists: Map<String, Long>, ticketid: String): CheckResult
    @Throws(CheckException::class)
    fun search(eventsAndCheckinLists: Map<String, Long>, query: String, page: Int): List<SearchResult>

    @Throws(CheckException::class)
    fun status(eventSlug: String, listId: Long): StatusResult?

    fun setSentry(sentry: SentryInterface)
}
